### PR TITLE
[Feature][Performance] Accelerate kNN on GPU

### PR DESCRIPTION
**Pull request by milesial**
## Description

When the input coordinates tensor is on the GPU, the current implementation of kNN does some GPU to CPU memory copies and CPU work with scipy. It also returns a graph on the CPU device, while we may expect the graph constructed from a GPU tensor to be on the GPU.

This PR accelerates kNN (topk algorithm) so that no such copies are done, and removes the need for scipy. The resulting implementation is much faster on GPU (up to x4 on the existing benchmark).

GPU benchmarks:

Baseline:
```
[100.00%] ··· api.bench_knn_graph.track_time                                                          
[100.00%] ··· ====== ===== ======================= ======================= =======================
              --                                              k                                   
              ------------ -----------------------------------------------------------------------
               size   dim             3                       5                       10          
              ====== ===== ======================= ======================= =======================
                50     16   0.0009597591997589916   0.0013467261498590233   0.0010055780501716071 
                50    128   0.0026513536500715417    0.002225380250092712   0.0013688179500604748 
                50    512    0.002554894849890843   0.0023144555496401155    0.002500285150017589 
               200     16   0.0042643542499718025    0.001552391750010429   0.0031550459996651627 
               200    128   0.0028832713498559313    0.003412479349935893    0.002601270599916461 
               200    512    0.002758960749997641   0.0032998006499838084    0.001834718450118089 
               1000    16    0.008994552950025537    0.007896198950038524    0.008799226400151383 
               1000   128    0.009393426949827699    0.00855530410008214     0.009593184149707668 
               1000   512    0.012859719499829225    0.01299530255018908     0.013523503150281613 
              ====== ===== ======================= ======================= =======================
```

This PR:
```
[100.00%] ··· ====== ===== ======================= ======================= =======================
              --                                              k                                    
              ------------ -----------------------------------------------------------------------
               size   dim             3                       5                       10          
              ====== ===== ======================= ======================= =======================
                50     16   0.0010200448036193847    0.000964303970336914   0.0010051936149597168 
                50    128   0.0009633088111877441    0.000966267204284668   0.0011321552276611327 
                50    512   0.0010028415679931642   0.0009913647651672362   0.0009624784469604491 
               200     16   0.0013418959617614748   0.0009911664009094239   0.0014054911613464356 
               200    128   0.0009869199752807617   0.0009781328201293946   0.0010408703804016112 
               200    512   0.0010021023750305176   0.0010544095993041992   0.0009980303764343262 
               1000    16   0.0022131391525268555   0.0022948911666870117    0.002367977523803711 
               1000   128    0.002474928092956543   0.0025180368423461915    0.002591118431091309 
               1000   512    0.003018016052246094   0.0031817327499389647    0.003524697494506836 
              ====== ===== ======================= ======================= =======================
```
One more thing that the benchmark does not account for is the time taken to copy the graph to GPU after the call to the old kNN. With this PR, this copy is not needed as the graph lives on the GPU.

As expected, when the input tensor is on the CPU, not much is changed:

Baseline:
```
[100.00%] ··· api.bench_knn_graph.track_time                                                         
[100.00%] ··· ====== ===== ======================= ======================= =======================
              --                                              k                                   
              ------------ -----------------------------------------------------------------------
               size   dim             3                       5                       10          
              ====== ===== ======================= ======================= =======================
                50     16   0.0012390059000608744    0.001562561350147007    0.001960483050061157 
                50    128    0.001852072549809236   0.0018715362999500939   0.0022495730001537593 
                50    512   0.0011760865501855733    0.00380662770003255    0.0021545202998822788 
               200     16   0.0023875225000665524   0.0027236581001488956    0.002691628000320634 
               200    128   0.0028037832998961676    0.002765762749913847    0.002806270050132298 
               200    512    0.002843265499905101    0.005400489100065897    0.007526781699925778 
               1000    16    0.013625155300178448    0.012653718649744405    0.013897375799933798 
               1000   128    0.016132800099876478    0.01622283674987557     0.01630912754990277  
               1000   512    0.02761213600024348     0.02796360355023353     0.02834183769991796  
              ====== ===== ======================= ======================= =======================
```
This PR:
```
[100.00%] ··· api.bench_knn_graph.track_time                                                        
[100.00%] ··· ====== ===== ======================= ======================= =======================
              --                                              k                                   
              ------------ -----------------------------------------------------------------------
               size   dim             3                       5                       10          
              ====== ===== ======================= ======================= =======================
                50     16   0.0012816177499189508   0.0009619387998100138   0.0010994348998792703 
                50    128   0.0010961471998598427   0.0016194106501643546   0.0019742950000363633 
                50    512   0.0017854944497230462   0.0013818412498949328   0.0013913328999478835 
               200     16   0.0021616040001390504   0.0022261916998104424   0.0022008218002156354 
               200    128    0.002242082550219493   0.0026428558998304653   0.0025004491999425227 
               200    512   0.0024608783998701254   0.0025859590001346077    0.002691755900013959 
               1000    16    0.008847501600030228    0.012652003249968402    0.011381140500088805 
               1000   128    0.015439252750002196    0.014687134749692632    0.014867262350162492 
               1000   512    0.03061338679981418     0.02509530525021546     0.026609388450015103 
              ====== ===== ======================= ======================= =======================

```

## Checklist

- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change

## Changes
- Accelerated topk implementation of kNN when the input is a GPU tensor

### CODE REVIEW

**Comment by VoVAllen** [comment]
To trigger regression tests: 
 - `@dgl-bot run [instance-type] [which tests] [compare-with-branch]`; 
 For example: `@dgl-bot run g4dn.4xlarge all dmlc/master` or `@dgl-bot run c5.9xlarge kernel,api dmlc/master`

**Comment by milesial** [comment]
@BarclayII @jermainewang

File: python/dgl/transform.py [hunk 1]
```diff
@@ -170,21 +170,17 @@ def _knn_graph_topk(x, k):
...
-    adj = sparse.csr_matrix(
-        (F.asnumpy(F.zeros_like(dst) + 1), (F.asnumpy(dst), F.asnumpy(src))),
-        shape=(n_samples * n_points, n_samples * n_points))
-    return convert.from_scipy(adj)
+    # index offset for each sample
+    offset = F.arange(0, n_samples, ctx=ctx) * n_points
+    offset = F.unsqueeze(offset, 1)
+    src = F.reshape(k_indices, (n_samples, n_points * k))
+    src = F.unsqueeze(src, 0) + offset
+    dst = F.repeat(F.arange(0, n_points, ctx=ctx), k, dim=0)
```
Comment (VoVAllen) [inline python/dgl/transform.py:181]: Will this cause larger memory consumption in the process?

Comment (milesial) [reply python/dgl/transform.py:181]: Overall, the src and dst tensors are the same shape as before, so memory should be the same (but now on GPU if the input is on GPU).
And now we don't have the potential memory overhead of `sparse.csr_matrix` (I don't know if it was significant).

**Review by VoVAllen — APPROVED** [review]

**Comment by lygztq** [comment]
Could you also modify `_segmented_knn_graph_topk` function below? I think both of these two functions need this improvement. Also, @hetong007 I think we can modify docstrings of these functions now (currently, the output of these functions will be copied to CPU regardless of the context of input)

**Comment by milesial** [comment]
@lygztq I modified the segmented version.

For the docstring, is something like this ok? :

"
If using the 'topk' algorithm, the returned graph is on the same device as input :attr:`x`. Else, the returned graph is on CPU, regardless of the context of the input :attr:`x`.
"

**Comment by lygztq** [comment]
> @lygztq I modified the segmented version.
> 
> For the docstring, is something like this ok? :
> 
> "
> If using the 'topk' algorithm, the returned graph is on the same device as input :attr:`x`. Else, the returned graph is on CPU, regardless of the context of the input :attr:`x`.
> "

I will have another PR for other algorithms later, so this docstring looks good to me.

**Comment by milesial** [comment]
Docstrings updated

**Comment by milesial** [comment]
Can this be merged? @lygztq

**Comment by lygztq** [comment]
> Can this be merged? @lygztq

It looks good to me, what do you think @hetong007?

**Comment by hetong007** [comment]
@lygztq Ideally we'd like to have the output knn graph from all algorithms on the same context as the input tensor. Since you are working on a succeeding patch, please remember to fix it as well.

Other than that, LGTM

**Review by hetong007 — APPROVED** [review]

**Pull request merged**